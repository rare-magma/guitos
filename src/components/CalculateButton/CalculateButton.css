.dropdown,
.dropdown-item,
.dropdown-item:focus,
.dropdown-item.disabled,
.dropdown-item:disabled,
.dropdown-item:hover {
  background: var(--lightbgcolor);
  color: var(--textcolor);
}

.dropdown-item:focus,
.dropdown-item:focus-visible,
.dropdown-item:hover {
  background: var(--lightbgcolor);
  border: 1px solid var(--pink);
  box-shadow: 0 0 0;
  color: var(--textcolor);
}

.dropdown-menu {
  min-width: inherit;
  overflow-x: hidden !important;
}

.dropdown-menu,
.dropdown-menu.show {
  background: var(--lightbgcolor);
  border: 0 solid var(--comment);
  border-radius: 0.375rem;
  color: var(--textcolor);
}
