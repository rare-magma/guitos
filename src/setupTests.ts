// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import * as matchers from "@testing-library/jest-dom/matchers";
import { cleanup } from "@testing-library/react";
import { randomUUID } from "node:crypto";
import { createElement } from "react";
import { afterEach, expect, vi } from "vitest";
import { Budget } from "./components/Budget/Budget";
import { ItemForm } from "./components/ItemForm/ItemForm";
import * as AppBudgetContext from "./context/BudgetContext";
import * as AppConfigContext from "./context/ConfigContext";
import * as AppGeneralContext from "./context/GeneralContext";

window.crypto.randomUUID = randomUUID;
global.URL.createObjectURL = vi.fn();

vi.mock("crypto", () => ({
  randomUUID: () => "035c2de4-00a4-403c-8f0e-f81339be9a4e",
}));

// silence recharts ResponsiveContainer error
vi.mock("recharts", async (importOriginal) => {
  const originalModule = await importOriginal();
  // eslint-disable-next-line @typescript-eslint/no-unsafe-return
  return {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    //@ts-ignore
    ...originalModule,
    ResponsiveContainer: () => createElement("div"),
  };
});

// extends Vitest's expect method with methods from react-testing-library
expect.extend(matchers);

export const configContextSpy = vi.spyOn(AppConfigContext, "useConfig");
export const budgetContextSpy = vi.spyOn(AppBudgetContext, "useBudget");
export const generalContextSpy = vi.spyOn(
  AppGeneralContext,
  "useGeneralContext",
);

beforeEach(() => {
  budgetContextSpy.mockReturnValue(testBudgetContext);
  configContextSpy.mockReturnValue(testConfigContext);
  generalContextSpy.mockReturnValue(testGeneralContext);
});

// runs a cleanup after each test case (e.g. clearing jsdom)
afterEach(() => {
  cleanup();
});

Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: vi.fn().mockImplementation((query: unknown) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(), // deprecated
    removeListener: vi.fn(), // deprecated
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
});

export const testBudget: Budget = {
  id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [{ id: 1, name: "expense1", value: 10 }],
    total: 10,
  },
  incomes: {
    items: [{ id: 2, name: "income1", value: 100 }],
    total: 100,
  },
  stats: {
    available: 90,
    withGoal: 80,
    saved: 10,
    goal: 10,
    reserves: 200,
  },
};
export const testBudgetClone: Budget = {
  ...testBudget,
  name: "2023-03-clone",
};

export const testBudget2: Budget = {
  id: "135b2ce4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-04",
  expenses: {
    items: [{ id: 1, name: "name", value: 50 }],
    total: 50,
  },
  incomes: {
    items: [{ id: 2, name: "name2", value: 200 }],
    total: 200,
  },
  stats: {
    available: 150,
    withGoal: 130,
    saved: 20,
    goal: 35,
    reserves: 30,
  },
};

export const testBigBudget: Budget = {
  id: "225c2de5-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [
      { id: 1, name: "name", value: 11378.64 },
      { id: 4, name: "name2", value: 11378.64 },
    ],
    total: 22757.28,
  },
  incomes: {
    items: [
      { id: 2, name: "name", value: 100.03 },
      { id: 3, name: "name2", value: 342783.83 },
    ],
    total: 342883.86,
  },
  stats: {
    available: 320126.58,
    withGoal: 148684.65,
    saved: 171441.93,
    goal: 50,
    reserves: 200,
  },
};

export const testJSONErrorBudget = `{
  id: "03123AAA5c2de4-00a4-403c-8f0e-f81339be9a4e",
  na2me: "2023-03",
  expens3es: {
    items: [{ id: "infinity", name: -1, value: "r" }],
    total: 10,
  },
  stats: {
    available: 0,
    withGoal: 0,
    saved: 0,
    goal: 10,
    reserves: 0,
  },
}`;

export const testCsv = `type,name,value
expense,rent,1000.00
expense,food,200.00
income,salary,2000.00
income,sale,100
goal,goal,10
reserves,reserves,0
`;

export const testCsvError = `type,name,value
expe2nse,rent,1000.00
expense,food,200.00,123,4
incomae,salary,2000.00
income,sale,100
goal,123,goal
goal,,goal,,,
reservaes,reserves,0
`;

export const testBudgetCsv = {
  id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
  name: "2023-03",
  expenses: {
    items: [
      new ItemForm({ id: 0, name: "rent", value: 1000 }),
      new ItemForm({ id: 1, name: "food", value: 200 }),
    ],
    total: 1200,
  },
  incomes: {
    items: [
      new ItemForm({ id: 2, name: "salary", value: 2000 }),
      new ItemForm({ id: 3, name: "sale", value: 100 }),
    ],
    total: 2100,
  },
  stats: {
    available: 900,
    withGoal: 690,
    saved: 210,
    goal: 10,
    reserves: 0,
  },
};

export const budgetNameList = [
  {
    id: "035c2de4-01a4-403c-8f0e-f81340be9a4e",
    name: "2023-03",
  },
  {
    id: "035c2de4-00a4-403c-8f0e-f81339be9a4e",
    name: "2023-04",
  },
  {
    id: "036c2de4-00a4-402c-8f0e-f81339be9a4e",
    name: "2023-05",
  },
];

export const itemForm1 = new ItemForm({
  id: 1,
  name: "name1",
  value: 10,
});

export const itemForm2 = new ItemForm({
  id: 2,
  name: "name2",
  value: 100,
});

export const testCalcHist = [
  {
    id: `${testBudget.id}-Expenses-1`,
    itemForm: itemForm1,
    changeValue: 123,
    operation: "add",
  },
  {
    id: `${testBudget.id}-Expenses-1`,
    itemForm: {
      id: 1,
      name: itemForm1.name,
      value: 133,
    },
    changeValue: 3,
    operation: "add",
  },
];

export const testIntlConfig = { locale: "en-US", currency: "USD" };
export const testSpanishIntlConfig = { locale: "es-ES", currency: "EUR" };

export const testBudgetList = [testBudget, testBudget2, testBigBudget];

export const testBudgetNameList = [
  {
    id: testBudget.id,
    item: "",
    name: testBudget.name,
  },
  {
    id: testBudget2.id,
    item: "",
    name: testBudget2.name,
  },
];

export const setIntlConfigMock = vi.fn();
export const handleCurrencyMock = vi.fn();
export const testConfigContext = {
  intlConfig: testIntlConfig || undefined,
  setIntlConfig: setIntlConfigMock,
  currency: "USD",
  handleCurrency: handleCurrencyMock,
};

export const testSpanishConfigContext = {
  intlConfig: testSpanishIntlConfig,
  setIntlConfig: setIntlConfigMock,
  currency: "EUR",
  handleCurrency: handleCurrencyMock,
};

export const setBudgetMock = vi.fn();
export const setBudgetListMock = vi.fn();
export const setBudgetNameListMock = vi.fn();
export const undoMock = vi.fn();
export const redoMock = vi.fn();
export const testEmptyBudgetContext = {
  budget: undefined,
  setBudget: setBudgetMock,

  budgetList: [],
  setBudgetList: setBudgetListMock,

  budgetNameList: [],
  setBudgetNameList: setBudgetNameListMock,

  revenuePercentage: 0,
  past: [],
  future: [],
  needReload: true,
  undo: undoMock,
  redo: redoMock,
  canUndo: false,
  canRedo: false,
};

export const testBudgetContext = {
  budget: testBudget,
  setBudget: setBudgetMock,

  budgetList: testBudgetList,
  setBudgetList: setBudgetListMock,

  budgetNameList: testBudgetNameList,
  setBudgetNameList: setBudgetNameListMock,

  revenuePercentage: 10,
  past: [],
  future: [],
  undo: undoMock,
  redo: redoMock,
  canUndo: false,
  canRedo: false,
};

export const setNeedReloadMock = vi.fn();
export const setLoadingFromDBMock = vi.fn();
export const setErrorMock = vi.fn();
export const setCsvErrorsMock = vi.fn();
export const setJsonErrorsMock = vi.fn();
export const setShowErrorMock = vi.fn();
export const setNotificationsMock = vi.fn();
export const handleErrorMock = vi.fn();
export const testGeneralContext = {
  needReload: false,
  setNeedReload: setNeedReloadMock,
  loadingFromDB: false,
  setLoadingFromDB: setLoadingFromDBMock,
  error: "",
  setError: setErrorMock,
  csvErrors: [],
  setCsvErrors: setCsvErrorsMock,
  jsonErrors: [],
  setJsonErrors: setJsonErrorsMock,
  showError: false,
  setShowError: setShowErrorMock,
  notifications: [],
  setNotifications: setNotificationsMock,
  handleError: handleErrorMock,
};

const jsonErrors: AppGeneralContext.JsonError[] = [
  {
    errors:
      "SyntaxError: Expected ',' or '}' after property value in JSON at position 209",
    file: "123.json",
  },
];

const csvErrors: AppGeneralContext.CsvError[] = [
  {
    errors: [
      {
        type: "FieldMismatch",
        code: "TooFewFields",
        message: "Line 0: Too few fields: expected 3 fields but parsed 2",
        row: 0,
      },
    ],
    file: "123.csv",
  },
];

const error = "Thrown error";

export const testErrorGeneralContext = {
  ...testGeneralContext,
  error: error,
  showError: true,
};

export const testJsonErrorGeneralContext = {
  ...testGeneralContext,
  jsonErrors: jsonErrors,
  showError: true,
};

export const testCsvErrorGeneralContext = {
  ...testGeneralContext,
  csvErrors: csvErrors,
  showError: true,
};
