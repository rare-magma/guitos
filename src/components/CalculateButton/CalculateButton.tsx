import { useEffect, useRef, useState } from "react";
import {
  Button,
  Dropdown,
  InputGroup,
  OverlayTrigger,
  Popover,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import {
  BsCheckLg,
  BsClockHistory,
  BsDashLg,
  BsPlusSlashMinus,
  BsXLg,
} from "react-icons/bs";
import { CgMathDivide, CgMathPlus } from "react-icons/cg";
import { useBudget } from "../../context/BudgetContext";
import { useConfig } from "../../context/ConfigContext";
import { useDB } from "../../hooks/useDB";
import { ItemForm } from "../ItemForm/ItemForm";
import "./CalculateButton.css";

interface CalculateButtonProps {
  itemForm: ItemForm;
  label: string;
  onCalculate: (changeValue: number, operation: ItemOperation) => void;
}

export type ItemOperation =
  | "name"
  | "value"
  | "add"
  | "subtract"
  | "multiply"
  | "divide";

export interface CalculationHistoryItem {
  id: string;
  itemForm: ItemForm;
  changeValue: number | undefined;
  operation: ItemOperation;
}

export function CalculateButton({
  itemForm,
  label,
  onCalculate,
}: CalculateButtonProps) {
  const [operation, setOperation] = useState<ItemOperation>("add");
  const [changeValue, setChangeValue] = useState(0);
  const [showHistory, setShowHistory] = useState(false);
  const [history, setHistory] = useState<CalculationHistoryItem[]>([]);
  const opButtonRef = useRef<HTMLButtonElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const { intlConfig } = useConfig();
  const { getCalcHist } = useDB();
  const { budget } = useBudget();
  const calcHistID = `${budget?.id}-${label}-${itemForm.id}`;

  function handleKeyPress(e: { key: string }) {
    if (e.key === "Enter") {
      handleCalculate();
      if (changeValue > 0) {
        opButtonRef?.current?.click();
      }
      setChangeValue(0);
    }
  }

  function handleCalculate() {
    if (changeValue > 0) {
      onCalculate(changeValue, operation);
      setShowHistory(false);
      getHistory();
    }
  }

  function handleHistory() {
    getHistory();
    setShowHistory(!showHistory);
  }

  function getHistory() {
    getCalcHist(calcHistID)
      .then((h) => setHistory(h))
      .catch((e: unknown) => {
        throw e;
      });
  }

  useEffect(() => {
    getHistory();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [showHistory]);

  return (
    <>
      <OverlayTrigger
        trigger="click"
        key={`${itemForm.id}-${label}-calculate-button-overlay-trigger`}
        placement="top"
        rootClose={true}
        overlay={
          <Popover id={`popover-calculate-button`}>
            <Popover.Body>
              <InputGroup
                size="sm"
                className="mb-1"
                key={`${itemForm.id}-${label}-operation-group`}
              >
                <Button
                  id={`item-${itemForm.id}-operation-history-button`}
                  key={`${itemForm.id}-${label}-operation-history-button`}
                  aria-label={"open operation history"}
                  variant="outline-secondary"
                  disabled={history.length > 0 ? false : true}
                  type="button"
                  onClick={handleHistory}
                >
                  <BsClockHistory aria-hidden />
                </Button>
                <Dropdown>
                  <Dropdown.Toggle
                    aria-label={"select type of operation on item value"}
                    aria-haspopup="true"
                    variant="outline-secondary"
                    id="dropdown-operation"
                  >
                    {operation === "add" && <CgMathPlus aria-hidden />}
                    {operation === "subtract" && <BsDashLg aria-hidden />}
                    {operation === "multiply" && <BsXLg aria-hidden />}
                    {operation === "divide" && <CgMathDivide aria-hidden />}
                  </Dropdown.Toggle>

                  <Dropdown.Menu>
                    <Dropdown.Item
                      aria-label="addition"
                      onClick={() => setOperation("add")}
                    >
                      <CgMathPlus aria-hidden />
                    </Dropdown.Item>
                    <Dropdown.Item
                      aria-label="subtraction"
                      onClick={() => setOperation("subtract")}
                    >
                      <BsDashLg aria-hidden />
                    </Dropdown.Item>
                    <Dropdown.Item
                      aria-label="multiplication"
                      onClick={() => setOperation("multiply")}
                    >
                      <BsXLg aria-hidden />
                    </Dropdown.Item>
                    <Dropdown.Item
                      aria-label="division"
                      onClick={() => setOperation("divide")}
                    >
                      <CgMathDivide aria-hidden />
                    </Dropdown.Item>
                  </Dropdown.Menu>
                </Dropdown>
                <CurrencyInput
                  id={`${label}-${itemForm.id}-operation-value`}
                  key={`item-${itemForm.id}-${label}-operation-value`}
                  className="text-end form-control straight-corners fixed-width-font"
                  aria-label={`${operation}`}
                  name="item-operate-value"
                  intlConfig={intlConfig}
                  defaultValue={0}
                  allowNegativeValue={false}
                  maxLength={14}
                  onKeyUp={handleKeyPress}
                  ref={inputRef}
                  onValueChange={(value) =>
                    setChangeValue(isNaN(Number(value)) ? 0 : Number(value))
                  }
                />
                <Button
                  id={`item-${itemForm.id}-trigger-operation-button`}
                  key={`${itemForm.id}-${label}-trigger-operation-button`}
                  aria-label={"apply change to item value"}
                  variant="outline-secondary"
                  type="button"
                  onClick={() => {
                    handleCalculate();
                    opButtonRef?.current?.click();
                  }}
                >
                  <BsCheckLg aria-hidden />
                </Button>
              </InputGroup>
              {showHistory && (
                <div style={{ maxHeight: "30vh", overflow: "auto" }}>
                  {history
                    .filter((i) => i.operation !== "value")
                    .map((item, index) => (
                      <InputGroup
                        size="sm"
                        className="mb-1"
                        key={`${item.id}-history-group-${index}`}
                      >
                        <CurrencyInput
                          id={`${label}-${itemForm.id}-${index}-history-value`}
                          key={`item-${itemForm.id}-${index}-${label}-history-value`}
                          className="text-end form-control straight-corners fixed-width-font"
                          aria-label={"item history value"}
                          name="item-history-value"
                          intlConfig={intlConfig}
                          defaultValue={item.itemForm.value}
                          disabled={true}
                        />
                        <InputGroup.Text>
                          {item.operation === "add" && (
                            <CgMathPlus aria-hidden />
                          )}
                          {item.operation === "subtract" && (
                            <BsDashLg aria-hidden />
                          )}
                          {item.operation === "multiply" && (
                            <BsXLg aria-hidden />
                          )}
                          {item.operation === "divide" && (
                            <CgMathDivide aria-hidden />
                          )}
                        </InputGroup.Text>
                        <CurrencyInput
                          id={`${label}-${itemForm.id}-${index}-history-changeValue`}
                          key={`item-${itemForm.id}-${index}-${label}-history-changeValue`}
                          className="text-end form-control straight-corners fixed-width-font"
                          aria-label={"item history change value"}
                          name="item-history-change-value"
                          intlConfig={intlConfig}
                          defaultValue={item.changeValue}
                          disabled={true}
                        />
                      </InputGroup>
                    ))
                    .reverse()}
                </div>
              )}
            </Popover.Body>
          </Popover>
        }
      >
        <Button
          id={`${label}-${itemForm.id}-operate-button`}
          key={`${itemForm.id}-${label}-operate-button`}
          aria-label={"select operation type to item value"}
          aria-haspopup="dialog"
          variant="outline-secondary"
          type="button"
          ref={opButtonRef}
          onClick={() => {
            setTimeout(() => {
              inputRef.current?.focus();
            }, 0);
          }}
        >
          <BsPlusSlashMinus aria-hidden />
        </Button>
      </OverlayTrigger>
    </>
  );
}
