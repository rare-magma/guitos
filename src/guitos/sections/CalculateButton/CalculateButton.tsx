import { useBudget } from "@guitos/context/BudgetContext";
import { useConfig } from "@guitos/context/ConfigContext";
import type { BudgetItem } from "@guitos/domain/budgetItem";
import type {
  CalculationHistoryItem,
  ItemOperation,
} from "@guitos/domain/calculationHistoryItem";
import { useDB } from "@guitos/hooks/useDB";
import { useCallback, useEffect, useId, useRef, useState } from "react";
import {
  Button,
  Dropdown,
  InputGroup,
  OverlayTrigger,
  Popover,
} from "react-bootstrap";
import CurrencyInput from "react-currency-input-field";
import {
  BsCheckLg,
  BsClockHistory,
  BsDashLg,
  BsPlusSlashMinus,
  BsXLg,
} from "react-icons/bs";
import { CgMathDivide, CgMathPlus } from "react-icons/cg";
import "./CalculateButton.css";

interface CalculateButtonProps {
  itemForm: BudgetItem;
  label: string;
  onCalculate: (changeValue: number, operation: ItemOperation) => void;
}

export function CalculateButton({
  itemForm,
  label,
  onCalculate,
}: CalculateButtonProps) {
  const [operation, setOperation] = useState<ItemOperation>("add");
  const [changeValue, setChangeValue] = useState(0);
  const [showHistory, setShowHistory] = useState(false);
  const [history, setHistory] = useState<CalculationHistoryItem[]>([]);
  const opButtonRef = useRef<HTMLButtonElement>(null);
  const inputRef = useRef<HTMLInputElement>(null);
  const { intlConfig } = useConfig();
  const { getCalcHist } = useDB();
  const { budget } = useBudget();
  const calcHistID = `${budget?.id}-${label}-${itemForm.id}`;

  function handleKeyPress(e: { key: string }) {
    if (e.key === "Enter") {
      handleCalculate();
      if (changeValue > 0) {
        opButtonRef?.current?.click();
      }
      setChangeValue(0);
    }
  }

  function handleCalculate() {
    if (changeValue > 0) {
      onCalculate(changeValue, operation);
      setShowHistory(false);
      getHistory();
    }
  }

  function handleHistory() {
    getHistory();
    setShowHistory(!showHistory);
  }

  const getHistory = useCallback(() => {
    getCalcHist(calcHistID)
      .then((h) => h && setHistory(h))
      .catch((e: unknown) => {
        throw e;
      });
  }, [calcHistID, getCalcHist]);

  useEffect(() => void getHistory(), [getHistory]);

  return (
    <OverlayTrigger
      trigger="click"
      key={`${itemForm.id}-${label}-calculate-button-overlay-trigger`}
      placement="top"
      rootClose={true}
      overlay={
        <Popover id={"popover-calculate-button"}>
          <Popover.Body>
            <InputGroup
              size="sm"
              className="mb-1"
              key={`${itemForm.id}-${label}-operation-group`}
            >
              <Button
                id={`item-${itemForm.id}-operation-history-button`}
                key={`${itemForm.id}-${label}-operation-history-button`}
                aria-label={"open operation history"}
                variant="outline-secondary"
                disabled={!(history.length > 0)}
                type="button"
                onClick={handleHistory}
              >
                <BsClockHistory aria-hidden={true} />
              </Button>
              <Dropdown>
                <Dropdown.Toggle
                  aria-label={"select type of operation on item value"}
                  aria-haspopup="true"
                  variant="outline-secondary"
                  id={useId()}
                >
                  {operation === "add" && <CgMathPlus aria-hidden={true} />}
                  {operation === "subtract" && <BsDashLg aria-hidden={true} />}
                  {operation === "multiply" && <BsXLg aria-hidden={true} />}
                  {operation === "divide" && (
                    <CgMathDivide aria-hidden={true} />
                  )}
                </Dropdown.Toggle>

                <Dropdown.Menu>
                  <Dropdown.Item
                    aria-label="addition"
                    onClick={() => setOperation("add")}
                  >
                    <CgMathPlus aria-hidden={true} />
                  </Dropdown.Item>
                  <Dropdown.Item
                    aria-label="subtraction"
                    onClick={() => setOperation("subtract")}
                  >
                    <BsDashLg aria-hidden={true} />
                  </Dropdown.Item>
                  <Dropdown.Item
                    aria-label="multiplication"
                    onClick={() => setOperation("multiply")}
                  >
                    <BsXLg aria-hidden={true} />
                  </Dropdown.Item>
                  <Dropdown.Item
                    aria-label="division"
                    onClick={() => setOperation("divide")}
                  >
                    <CgMathDivide aria-hidden={true} />
                  </Dropdown.Item>
                </Dropdown.Menu>
              </Dropdown>
              <CurrencyInput
                id={`${label}-${itemForm.id}-operation-value`}
                key={`item-${itemForm.id}-${label}-operation-value`}
                className="text-end form-control straight-corners fixed-width-font"
                aria-label={`${operation}`}
                name="item-operate-value"
                intlConfig={intlConfig}
                defaultValue={0}
                allowNegativeValue={false}
                maxLength={14}
                onKeyUp={handleKeyPress}
                ref={inputRef}
                onValueChange={(value) =>
                  setChangeValue(
                    Number.isNaN(Number(value)) ? 0 : Number(value),
                  )
                }
              />
              <Button
                id={`item-${itemForm.id}-trigger-operation-button`}
                key={`${itemForm.id}-${label}-trigger-operation-button`}
                aria-label={"apply change to item value"}
                variant="outline-secondary"
                type="button"
                onClick={() => {
                  handleCalculate();
                  opButtonRef?.current?.click();
                }}
              >
                <BsCheckLg aria-hidden={true} />
              </Button>
            </InputGroup>
            {showHistory && (
              <div style={{ maxHeight: "30vh", overflow: "auto" }}>
                {history
                  .filter((i) => i.operation !== "value")
                  .map((item, index) => (
                    <InputGroup
                      size="sm"
                      className="mb-1"
                      key={`${item.id}-history-group-${index}`}
                    >
                      <CurrencyInput
                        id={`${label}-${itemForm.id}-${index}-history-value`}
                        key={`item-${itemForm.id}-${index}-${label}-history-value`}
                        className="text-end form-control straight-corners fixed-width-font"
                        aria-label={"item history value"}
                        name="item-history-value"
                        intlConfig={intlConfig}
                        defaultValue={item.itemForm.value}
                        disabled={true}
                      />
                      <InputGroup.Text>
                        {item.operation === "add" && (
                          <CgMathPlus aria-hidden={true} />
                        )}
                        {item.operation === "subtract" && (
                          <BsDashLg aria-hidden={true} />
                        )}
                        {item.operation === "multiply" && (
                          <BsXLg aria-hidden={true} />
                        )}
                        {item.operation === "divide" && (
                          <CgMathDivide aria-hidden={true} />
                        )}
                      </InputGroup.Text>
                      <CurrencyInput
                        id={`${label}-${itemForm.id}-${index}-history-changeValue`}
                        key={`item-${itemForm.id}-${index}-${label}-history-changeValue`}
                        className="text-end form-control straight-corners fixed-width-font"
                        aria-label={"item history change value"}
                        name="item-history-change-value"
                        intlConfig={intlConfig}
                        defaultValue={item.changeValue}
                        disabled={true}
                      />
                    </InputGroup>
                  ))
                  .toReversed()}
              </div>
            )}
          </Popover.Body>
        </Popover>
      }
    >
      <Button
        id={`${label}-${itemForm.id}-operate-button`}
        key={`${itemForm.id}-${label}-operate-button`}
        aria-label={"select operation type to item value"}
        aria-haspopup="dialog"
        variant="outline-secondary"
        type="button"
        ref={opButtonRef}
        onClick={() => {
          setTimeout(() => {
            inputRef.current?.focus();
          }, 0);
          getHistory();
        }}
      >
        <BsPlusSlashMinus aria-hidden={true} />
      </Button>
    </OverlayTrigger>
  );
}
