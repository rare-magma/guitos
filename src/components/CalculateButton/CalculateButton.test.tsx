import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { vi } from "vitest";
import { itemForm1 } from "../../setupTests";
import CalculateButton from "./CalculateButton";

describe("CalculateButton", () => {
  const onCalculate = vi.fn();
  const comp = (
    <CalculateButton
      itemForm={itemForm1}
      label="Expense"
      onCalculate={onCalculate}
    />
  );

  beforeEach(() => {
    render(comp);
  });

  it("matches snapshot", () => {
    expect(comp).toMatchSnapshot();
  });

  it("renders initial state", () => {
    expect(
      screen.getByLabelText("select operation type to item value"),
    ).toBeInTheDocument();
  });

  it("opens popover when clicking the button", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);

    expect(
      screen.getByRole("button", {
        name: "select type of operation on item value",
      }),
    ).toBeInTheDocument();

    expect(screen.getByLabelText("addition")).toBeInTheDocument();

    expect(
      screen.getByRole("button", {
        name: "apply change to item value",
      }),
    ).toBeInTheDocument();
  });

  it("closes when clicking the button", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);

    const button2 = screen.getByRole("button", {
      name: "select type of operation on item value",
    });

    await userEvent.click(button);

    expect(button2).not.toBeInTheDocument();
  });

  it("closes when pressing Escape key", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);

    const button2 = screen.getByRole("button", {
      name: "select type of operation on item value",
    });

    await userEvent.type(screen.getByLabelText("addition"), "{Escape}");

    expect(button2).not.toBeInTheDocument();
  });

  it("calls onCalculate when accepting change > 0", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);
    const acceptButton = screen.getByRole("button", {
      name: "apply change to item value",
    });

    await userEvent.type(screen.getByLabelText("add"), "123");
    await userEvent.click(acceptButton);

    expect(onCalculate).toBeCalledWith(123, "add");
  });

  it("calls onCalculate when change > 0 and enter is pressed", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);
    await userEvent.type(screen.getByLabelText("addition"), "123");
    await userEvent.type(screen.getByLabelText("addition"), "{enter}");

    expect(onCalculate).toBeCalledWith(123, "add");
  });

  it("calls onCalculate with sub", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);
    const acceptButton = screen.getByRole("button", {
      name: "apply change to item value",
    });

    await userEvent.type(screen.getByLabelText("add"), "123");

    await userEvent.click(screen.getByLabelText("subtraction"));
    await userEvent.click(acceptButton);

    expect(onCalculate).toBeCalledWith(123, "subtract");
  });

  it("calls onCalculate with multiply", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);
    const acceptButton = screen.getByRole("button", {
      name: "apply change to item value",
    });

    await userEvent.type(screen.getByLabelText("add"), "123");

    await userEvent.click(screen.getByLabelText("multiplication"));
    await userEvent.click(acceptButton);

    expect(onCalculate).toBeCalledWith(123, "multiply");
  });

  it("calls onCalculate with div", async () => {
    const button = screen.getByRole("button", {
      name: "select operation type to item value",
    });
    await userEvent.click(button);
    const acceptButton = screen.getByRole("button", {
      name: "apply change to item value",
    });

    await userEvent.type(screen.getByLabelText("add"), "123");

    await userEvent.click(screen.getByLabelText("division"));
    await userEvent.click(acceptButton);

    expect(onCalculate).toBeCalledWith(123, "divide");
  });
});
