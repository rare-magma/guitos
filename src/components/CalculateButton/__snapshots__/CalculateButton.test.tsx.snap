// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalculateButton > matches snapshot 1`] = `
<CalculateButton
  itemForm={
    ItemForm {
      "id": 1,
      "name": "name1",
      "value": 10,
    }
  }
  label="Expense"
  onCalculate={[MockFunction spy]}
/>
`;
